### DISCUSSION TITLE
How to actually use the AutoCompleteBox?

### DISCUSSION THREAD

**Discussion by argon522** [AvaloniaUI/Avalonia#17583] [Q&A] [answered]
I'm new to Avalonia and trying to figure things out. Currently I am trying to set up an auto-complete box control where the item selection is bound to an observable collection of objects.  Sounds like it should be easy, right?  turns out it doesn't seem to be.  I have a semi working method using a combo box, but would prefer the user be able to type in the selection vs having to switch to the mouse, then back to typing for form fill in.

xaml for the working combobox:

```
<ComboBox ItemsSource="{Binding MoveType}" SelectedItem="{Binding SelectedMovement}" >

	<ComboBox.ItemTemplate>

		<DataTemplate>

			<TextBlock Text="{Binding Movement}" ToolTip.Tip="{Binding Description}"/>

		</DataTemplate>

	</ComboBox.ItemTemplate>

</ComboBox>
```

In a perfect world I would expect to be able to just switch around some names and the AutoCompleteBox would be an essentially drop-in replacement.  Apparently, I am wrong and the autocomplete box seems to have significantly different behavior to the Combo box.

My current kinda not working autocomplete xaml:

```
<AutoCompleteBox ItemsSource="{Binding MoveType}"
 
				 IsTextCompletionEnabled="True" 

				 SelectedItem="{Binding SelectedItem}"

				 FilterMode="Contains" x:CompileBindings="False">

	<AutoCompleteBox.ItemTemplate>

		<DataTemplate>

			<TextBlock Text="{Binding Movement}" ToolTip.Tip="{Binding Description}"/>

		</DataTemplate>

	</AutoCompleteBox.ItemTemplate>

</AutoCompleteBox>
```

This code only kind of works.  I get a proper dropdown showing the correct information.  That is what works.
Searching is non-functional, as it searches against the stringified version of the object name.  I have been unable to find a way through the sparse documentation to change it to search against a property of an object. Same with displaying the selected item.  I can select the correct item if I get the dropdown to show, but it just displays the stringified name of the object with yet again no way have it display an object property.   This is in contrast to the ComboBox, that just works like one would expect, with the exception of not being able to type in anything.  You will also note that I had to add x:CompileBindings="False" to the autocomplete box, as compiled bindings in the data template were failing to compile, yet it works fine with them disabled. Unlike the ComboBox, which once again, just works like it should.

I don't know if I am just doing something horribly wrong, or if the autocomplete box is just not a complete feature, and I need to look into other options.  Any help would be appreciated.

edited to try and clean up code presentation

**Comment by timunie** [comment]
Would override ToString in your ItemViewModel be a solution for you?

**Reply by argon522** [reply]
Overriding ToString in the Model did work.  Thank you.  

I personally believe that things like that should be handled in the view, as the combo box appears to do, as overriding a method in the model to fix a display issue goes against the separation given by MVVM.  But for now that can work as a quick fix.

**Reply by timunie** [reply]
I think the ViewModel (or ItemViewModel) is allowed to do that, but the Model probably not. Anyways, could be also useful for debugging having a meaningfol ToString override.

**Comment by tobyfirth** [comment, marked as answer]
The autocompletebox looks like it searches on the result of calling ToString on the objects in the ItemSource. Looking at the example in the ControlCatalog sample it look like you can either use the property [ValueMemberBinding](https://github.com/AvaloniaUI/Avalonia/blob/ed0cc758e4c1bae1938fbfed738b254627977a59/samples/ControlCatalog/Pages/AutoCompleteBoxPage.xaml#L51) to bind to a specific member of the object or create a [value converter](https://github.com/AvaloniaUI/Avalonia/blob/ed0cc758e4c1bae1938fbfed738b254627977a59/samples/ControlCatalog/Pages/AutoCompleteBoxPage.xaml.cs#L120) which allows you to return any string you want from the binding.

**Reply by argon522** [reply]
ValueMemberBinding appears to be exactly what I was looking for.  Thank you!
